src/common/allocator.h

```diff
@@ -0,0 +1,65 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_COMMON_ALLOCATOR_H
+#define CEPH_COMMON_ALLOCATOR_H
+
+#include <malloc.h>
+#include <memory>
+#include "acconfig.h"
+
+namespace ceph {
+
+#ifdef LIBTCMALLOC_MISSING_ALIGNED_ALLOC
+
+// If libtcmalloc is missing 'aligned_alloc', provide a new allocator class that
+// uses memalign which is what newer versions of tcmalloc do internally. C++17
+// will automatically use 'operator new(size_t, align_val_t)' for aligned
+// structures, which will invoke the missing 'aligned_alloc' tcmalloc function.
+// This method was added to tcmalloc (gperftools) in commit d406f228 after
+// the 2.6.1 release was tagged.
+template <typename T>
+struct allocator : public std::allocator<T> {
+  using pointer = typename std::allocator<T>::pointer;
+  using size_type = typename std::allocator<T>::size_type;
+
+  template<class U>
+  struct rebind {
+    typedef allocator<U> other;
+  };
+
+  allocator() noexcept {
+  }
+
+  allocator(const allocator& other) noexcept : std::allocator<T>(other) {
+  }
+
+  template <class U>
+  allocator(const allocator<U>& other) noexcept : std::allocator<T>(other) {
+  }
+
+  pointer allocate(size_type n, const void* hint = nullptr) {
+    if (n > this->max_size()) {
+      throw std::bad_alloc();
+    }
+
+    if (alignof(T) > 1) {
```

Comment:
could use
```c++
if constexpr alignof(T) > __STDCPP_DEFAULT_NEW_ALIGNMENT__) {
// ...
}
```

Reply:
I didn’t want to use a potentially private macro. Is that actually part of the standard? The “__” makes me think it isn’t.

Reply:
yeah, it is. see http://eel.is/c++draft/cpp.predefined#1.6 and https://en.cppreference.com/w/cpp/preprocessor/replace . it was added by http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0035r4.html .

Reply:
updated

Reply:
could use `constexpr if` to enforce the compile-time condition.

Reply:
GCC 8 and 9 were both smart enough to not need it to optimize out the code.

Reply:
fair enough.